### ISSUE TITLE
`--verifytypes` is reporting overloads with ellipsis default values as functions without default param

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2689]
**Describe the bug**
Given the following code (forgive the contrived overloads):

```python
from typing import overload


class Bar:
    @overload
    def __init__(self, value: int, *, optional: str | None = ...) -> None:
        ...

    @overload
    def __init__(self, value: str, *, optional: str | None = ...) -> None:
        ...

    def __init__(self, value: int | str, *, optional: str | None = None) -> None:
        self.value: int | str = value
        self.optional: str | None = optional
```

Running pyright with `--verifytypes` reports 2 functions without default params.

**Expected behavior**
0 functions without default params should be reported.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.194

**Comment by erictraut** [comment]
Yes, this is intentional. The reason it's important to include default values in all overload signatures is for "signature help" functionality included in language servers like pylance. When the user types `Bar(`, they receive a popup that provides details about the signature. Users have repeatedly told us that they find the default values very useful here, but if the overload includes `...`, this is obscured. Do you see any problem with including the default value in the overload signatures?

**Comment by bryanforbes** [comment]
This kind of presents a conundrum: for stub files, the recommendation (and what typeshed does and enforces using `flake8-pyi`) is to use `...` for default values in overloads; however for Python modules, it looks like the recommendation (and what other typed projects like mypy do) is to code the actual default values in overloads. I'm not opposed to using actual default values, but coming from several stub projects it's odd to see this recommendation. Has there been a proposal to add default values to the overload signatures in typeshed?

At the very least, it may be good to have an option to ignore overload default values for `--verifytypes` to support both cases.

**Comment by erictraut** [comment]
The `flake8` and typeshed rules were established prior to the popularization of language servers, so they didn't take into account this important use case. 

Here's a [document](https://github.com/python/typing/blob/master/docs/source/libraries.rst#inlined-type-annotations-and-type-stubs) that provides guidance for library developers. This document is based on one that I initially authored, but it is now part of the python/typing repo as it applies more broadly than just pylance/pyright. This document includes the following sentence:

> In many existing type stubs (such as those found in typeshed), default parameter values are replaced with with ... and all docstrings are removed. We recommend that default values and docstrings remain within the type stub file so language servers can display this information to developers.

We haven't yet pushed for making this change broadly in typeshed, but we may do so in the future depending on how loud the feedback is from pylance users. So far, we haven't received that much feedback because pylance special-cases the stdlib modules (effectively enhancing the information provided in typeshed stub files) so it can provide docstrings and default argument values for these modules.

The `--verifytypes` option does not count a function as having an "unknown type" if it uses "..." as a default argument. (The same is true if a function or class omits a docstring.) And it doesn't generate a non-zero exit code if "..." default arguments are present. It does flag these, but as a library owner you can ignore them if you don't want to provide this information.

**Comment by bryanforbes** [comment]
Thanks for the info. I was unaware of the change in recommendations, so I'll go ahead and close this.

**Comment by JelleZijlstra** [comment]
For what it's worth I'd be supportive of changing typeshed to allow defaults and perhaps even docstrings. If we can make typeshed useful for more use cases without hurting typing, we should do it.

**Comment by erictraut** [comment]
Thanks @JelleZijlstra. Like I mentioned above, docstrings and default argument values are less important in stdlib stubs (at least for pylance, because it ships with enhancements that provide these details), but these are a bigger issue for third-party stubs. I think that typeshed's CI scripts currently reject docstrings and default arguments. Perhaps as a first step we could loosen that requirement for third-party stubs in typeshed? That would allow contributors to these stubs to include defaults if they desire.

**Comment by JelleZijlstra** [comment]
I'd be OK with that, but of course we'd also need buy-in from other maintainers.

A couple of other thoughts:
- We should ensure that all tools that consume stubs can deal with default arguments and docstrings. I know pytype has a separate parser specifically for stubs, for example, and perhaps it will just choke on docstrings.
- For defaults, we'll need to set some rules on what kinds of defaults are acceptable. Certainly True or False or string literals should be OK, but `default=some_function_call()` shouldn't be, but what about a reference to a global variable?
- We should enhance stubtest to verify that runtime and stub docstrings and defaults match.
- Should we also allow values for class variables and global constants?
